README.md
# Terraform Provider Okta

- [![Build Status](https://travis-ci.org/terraform-providers/terraform-provider-okta.svg?branch=master)](https://travis-ci.org/terraform-providers/terraform-provider-okta)
- Website: https://www.terraform.io
- [![Gitter chat](https://badges.gitter.im/hashicorp-terraform/Lobby.png)](https://gitter.im/hashicorp-terraform/Lobby)
- Mailing list: [Google Groups](http://groups.google.com/group/terraform-tool)

<img src="https://cdn.rawgit.com/hashicorp/terraform-website/master/content/source/assets/images/logo-hashicorp.svg" width="600px">

## Requirements

- [Terraform](https://www.terraform.io/downloads.html) 0.12.x
- [Go](https://golang.org/doc/install) 1.12 (to build the provider plugin)

## Usage

This plugin requires two inputs to run: the okta organization name and the okta api token. The okta base url is not required and will default to "okta.com" if left out.

You can specify the inputs in your tf plan:

```
provider "okta" {
  org_name  = <okta instance name, e.g. dev-XXXXXX>
  api_token = <okta instance api token with the Administrator role>
  base_url  = <okta base url, e.g. oktapreview.com>

  // Optional settings, https://en.wikipedia.org/wiki/Exponential_backoff
  max_retries      = <number of retries on api calls, default: 5>
  backoff          = <enable exponential backoff strategy for rate limits, default = true>
  min_wait_seconds = <min number of seconds to wait on backoff, default: 30>
  max_wait_seconds = <max number of seconds to wait on backoff, default: 300>
}
```

OR you can specify environment variables:

```
OKTA_ORG_NAME=<okta instance name, e.g. dev-XXXXXX>
OKTA_API_TOKEN=<okta instance api token with the Administrator role>
OKTA_BASE_URL=<okta base url, e.g. oktapreview.com>
```

## Examples

As we build out resources we build concomitant acceptance tests that require use to create resource config that actually creates and modifies real resources. We decided to put these test fixtures to good use and provide them [as examples here.](./examples)

## Building The Provider

Clone repository to: `$GOPATH/src/github.com/terraform-providers/terraform-provider-okta`

```sh
$ mkdir -p $GOPATH/src/github.com/terraform-providers; cd $GOPATH/src/github.com/terraform-providers
$ git clone git@github.com:terraform-providers/terraform-provider-okta
```

Enter the provider directory and build the provider. Ensure you have Go Modules enabled, depending on the version of Go you are using, you may have to flip it on with `GO111MODULE=on`.

```sh
$ cd $GOPATH/src/github.com/terraform-providers/terraform-provider-okta
$ make build
```

## Using the provider

Example terraform plan:

```
provider "okta" {
  org_name  = "dev-XXXXX"
  api_token = "XXXXXXXXXXXXXXXXXXXXXXXX"
  base_url  = "oktapreview.com"
}

resource "okta_user" "blah" {
  first_name = "blah"
  last_name  = "blergh"
  email      = "XXXXX@XXXXXXXX.XXX"
  login      = "XXXXX@XXXXXXXX.XXX"
}
```

## Developing the Provider

If you wish to work on the provider, you'll first need [Go](http://www.golang.org) installed on your machine (version 1.8+ is *required*). You'll also need to correctly setup a [GOPATH](http://golang.org/doc/code.html#GOPATH), as well as adding `$GOPATH/bin` to your `$PATH`.

To compile the provider, run `make build`. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.

```sh
$ make build
...
$ $GOPATH/bin/terraform-provider-okta
...
```

In order to test the provider, you can simply run `make test`. The acceptance tests require an API token and a corresponding Okta org, if you use dotenv, you can `cp .env.sample .env` and add your Okta settings there, and prefix make test with `dotenv`.

```sh
$ make test
```

In order to run the full suite of Acceptance tests, run `make testacc`.

*Note:* Acceptance tests create real resources, and often cost money to run.

```sh
$ make testacc
```


CHANGELOG.md
## 3.0.0

FEATURES:

* Updated provider to support Terraform v0.12.0

## 3.0.1

FEATURES:

* **New Resource:** `okta_inline_hook`

ENHANCEMENTS:

* Add missing okta_idp_saml settings

## 3.0.2

ENHANCEMENTS:

* Use backoff/retries functionality for XML API calls

## 3.0.3

FEATURES:

* **New Data Source:** okta_idp_saml

ENHANCEMENTS:

* Support import user by email

## 3.0.4

FEATURES:

* **New Data Source:** Add okta_app_saml data source
* **New Data Source:** Add okta_app_metadata_saml data source
* **New Data Source:** Add okta_idp_metadata_saml data source

ENHANCEMENTS:

* Change type of custom_profile_attributes from map to JSON string to support all types

BUG FIXES:

* Fix group filter bug, filter_type and filter_value were not being sync'd

## 3.0.5

BUG FIXES:

* Fix bug introduced in v3.0.4. User data source was not updated to the new caustom_profile_attribute type
* Added test to cover this scenario, tests were passiing

## 3.0.6

ENHANCEMENTS:

* Allow client_id to be set on OIDC application, while also maintaining the computed version. With some auth methods, such as basic auth, this is possible.

## 3.0.7

ENHANCEMENTS:

* Add group_assignments for SAML and social IdPs

## 3.0.8

ENHANCEMENTS:

* Add issuer_mode to social IdP. Our test org does not have a custom domain setup, thus it was working there but not in other orgs. Hard to test both scenarios in one org.

## 3.0.9

FEATURES:

* **New Resource:** `okta_template_email`
* **New Resource:** `okta_group_roles`

## 3.0.10

FEATURES:

* **New Resource:** `okta_network_zone`

## 3.0.11

BUG FIXES:

* Fix occasional panic when creating a user schema see https://github.com/articulate/terraform-provider-okta/issues/144
* Users in LOCKED_OUT state are unlocked when config is ACTIVE https://github.com/articulate/terraform-provider-okta/issues/225

## 3.0.12

BUG FIXES:

* Ensure schema does not panic after retry

## 3.0.13

FEATURES:

* **New Resource:** `okta_user_base_schema`

ENHANCEMENTS:

* Add missing attribute, match_type and match_attribute, on social idp resource

## 3.0.14

BUG FIXES:

* Fix logic around including/excluding networks on policy rules

## 3.0.15

ENHANCEMENTS:

* Update Okta SDK
* Filter out GROUP based admin roles when processing user `admin_roles` attribute

## 3.0.16

* Fix issues around `okta_policy_rule_idp_discovery`
    * `app_include` and `app_exlcude` were missing required properties
    * `user_identifier_type` was being added even when not defined, causing API errors
* Fix integer array type

## 3.0.17

FEATURES:

* **New Resource:** `okta_app_user_schema`
* **New Resource:** `okta_app_user_base_schema`
* **New Resource:** `okta_app_user` resource
* **New Resource:** `okta_app_group` resource

ENHANCEMENTS:

* Add `required` field to base schema

## 3.0.18

ENHANCEMENTS:

* Support SHA-1 signing algorithm on IdPs

BUG FIXES:

* Fix bug where audience is reset on IdP update because it is omitted from the payload

## 3.0.19

BUG FIXES:

* Fix diff issues around `okta_policy_rule_idp_discovery`
* Allow `provisioning_action` for IdPs to be set to `DISABLED`

## 3.0.20

BUG FIXES:

* Fix `okta_auth_server_claim`, `group_filter_type` could not be set to `STARTS_WITH` due to a typo

## 3.0.21

ENHANCEMENTS:

* Expose scope property on `okta_user_schema`
* Allow setting of OAuth application visibility settings

## 3.0.22

BUG FIXES:

* Send `profileMaster` along with IdP, so the config is recognized by Okta API
* Fix bug in SDK related to retries and the request body being empty on subsequent requests.

## 3.0.23

ENHANCEMENTS:

* Add `external_name` property to the `okta_app_user_schema` and `okta_user_schema`

## 3.0.24

ENHANCEMENTS:

* Support `profile` on `okta_oauth_app` resource

## 3.0.25

ENHANCEMENTS:

* Support setting an auth server scope as the default
* Support `profile` and `priority` on `okta_app_group_assignment`
* Support `profile` on `okta_app_user`

BUG FIXES:

* Fix bug with supporting `profile` on `okta_oauth_app` resource

## 3.0.26

ENHANCEMENTS:

* Support array enums in `okta_user_schema` and `okta_app_user_schema` as `array_enum` and `array_one_of`

## 3.0.27

ENHANCEMENTS:

* Update refresh token window validation to account for new upper limit of 5 years

## 3.0.28

BUG FIXES:

* Remove resource from state on 404. [GH-269]

## 3.0.29

BUG FIXES:

* Ensure we safely sync auth server properties. [GH-299]
* MANUAL rotation mode can only be set on an auth server on update. Ensure we run update after create for that scenario. [GH-287]

## 3.0.30

ENHANCEMENT:

* Update to new separate Terraform SDK [GH-307]

## 3.0.31

BUG FIXES:

* Ensure `okta_app_group_assignment` resource syncs using the right read function. [GH-307]

## 3.0.32

BUG FIXES:

* Ensure `okta_app_group_assignment` and `okta_app_user` resources properly take multiple ids on the import functions. [GH-307]
* Ensure `okta_user` does not error on 404 [GH-313]

## 3.0.33

FEATURES:

* **New Resource:** `okta_profile_mapping` [GH-246]
* **New Resource:** `okta_app_basic_auth` [GH-329]

## 3.0.34

BUG FIXES:

* Policy values could not be set to 0. Doing so resulted in the SDK omitting them, resulting in Okta resetting the values to default.

## 3.0.35

ENHANCEMENT:

* Require target_id on `okta_profile_mapping` to avoid ambiguity

FEATURES:

* **New Data Source:** `okta_user_profile_mapping_source` [GH-340]

## 3.0.36

BUG FIXES

* Schema merging helper function was mutating input schema causing side effects when used in a particular way. Used shallow copying to avoid this side effect. [GH-338] 
* Ensure response is not nil when checking status code [GH-307]

## 3.0.37

BUG FIXES

* Ensure `index` is sync'd on import to avoid recreation.

## 3.0.38

ENHANCEMENT:

* Support `password`, `recovery_answer`, and `recovery_question` as attributes on the `okta_user` resource.


okta/provider.go
// Package okta terraform configuration for an okta site
package okta

import (
	"fmt"
	"log"

	"gopkg.in/ini.v1"

	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
	"github.com/hashicorp/terraform-plugin-sdk/terraform"
)

// Resource names, defined in place, used throughout the provider and tests
const (
	appAutoLogin           = "okta_app_auto_login"
	appBookmark            = "okta_app_bookmark"
	appGroupAssignment     = "okta_app_group_assignment"
	appUser                = "okta_app_user"
	appOAuth               = "okta_app_oauth"
	appOAuthRedirectUri    = "okta_app_oauth_redirect_uri"
	appSaml                = "okta_app_saml"
	appSecurePasswordStore = "okta_app_secure_password_store"
	appSwa                 = "okta_app_swa"
	appThreeField          = "okta_app_three_field"
	appUserSchema          = "okta_app_user_schema"
	appUserBaseSchema      = "okta_app_user_base_schema"
	authServer             = "okta_auth_server"
	authServerClaim        = "okta_auth_server_claim"
	authServerPolicy       = "okta_auth_server_policy"
	authServerPolicyRule   = "okta_auth_server_policy_rule"
	authServerScope        = "okta_auth_server_scope"
	factor                 = "okta_factor"
	groupRoles             = "okta_group_roles"
	groupRule              = "okta_group_rule"
	identityProvider       = "okta_identity_provider"
	idpResource            = "okta_idp_oidc"
	idpSaml                = "okta_idp_saml"
	idpSamlKey             = "okta_idp_saml_key"
	idpSocial              = "okta_idp_social"
	inlineHook             = "okta_inline_hook"
	networkZone            = "okta_network_zone"
	oktaGroup              = "okta_group"
	oktaUser               = "okta_user"
	policyMfa              = "okta_policy_mfa"
	policyPassword         = "okta_policy_password"
	policyRuleIdpDiscovery = "okta_policy_rule_idp_discovery"
	policyRuleMfa          = "okta_policy_rule_mfa"
	policyRulePassword     = "okta_policy_rule_password"
	policyRuleSignOn       = "okta_policy_rule_signon"
	policySignOn           = "okta_policy_signon"
	templateEmail          = "okta_template_email"
	trustedOrigin          = "okta_trusted_origin"
	userBaseSchema         = "okta_user_base_schema"
	userSchema             = "okta_user_schema"
)

// Provider establishes a client connection to an okta site
// determined by its schema string values
func Provider() terraform.ResourceProvider {
	deprecatedPolicies := dataSourceDefaultPolicies()
	deprecatedPolicies.DeprecationMessage = "This data source will be deprecated in favor of okta_default_policy or okta_policy data sources."

	return &schema.Provider{
		Schema: map[string]*schema.Schema{
			"org_name": {
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("OKTA_ORG_NAME", nil),
				Description: "The organization to manage in Okta.",
			},
			"api_token": {
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("OKTA_API_TOKEN", nil),
				Description: "API Token granting privileges to Okta API.",
			},
			"base_url": {
				Type:        schema.TypeString,
				Optional:    true,
				DefaultFunc: schema.EnvDefaultFunc("OKTA_BASE_URL", "okta.com"),
				Description: "The Okta url. (Use 'oktapreview.com' for Okta testing)",
			},
			"config_file": {
				Type:        schema.TypeString,
				Optional:    true,
				Description: "Path to file containing Okta provider configuration. The file should be in YAML format and should use the same key names as used by the provider config in Terraform. The settings in the file will always take precedence over properties in config or environment.",
			},
			"backoff": {
				Type:        schema.TypeBool,
				Optional:    true,
				Default:     true,
				Description: "Use exponential back off strategy for rate limits.",
			},
			"min_wait_seconds": {
				Type:        schema.TypeInt,
				Optional:    true,
				Default:     30,
				Description: "minimum seconds to wait when rate limit is hit. We use exponential backoffs when backoff is enabled.",
			},
			"max_wait_seconds": {
				Type:        schema.TypeInt,
				Optional:    true,
				Default:     300,
				Description: "maximum seconds to wait when rate limit is hit. We use exponential backoffs when backoff is enabled.",
			},
			"max_retries": {
				Type:         schema.TypeInt,
				Optional:     true,
				Default:      5,
				ValidateFunc: validation.IntAtMost(100), // Have to cut it off somewhere right?
				Description:  "maximum number of retries to attempt before erroring out.",
			},
			"parallelism": {
				Type:        schema.TypeInt,
				Optional:    true,
				Default:     1,
				Description: "Number of concurrent requests to make within a resource where bulk operations are not possible. Take note of https://developer.okta.com/docs/api/getting_started/rate-limits.",
			},
		},

		ResourcesMap: map[string]*schema.Resource{
			appAutoLogin:           resourceAppAutoLogin(),
			appBookmark:            resourceAppBookmark(),
			appGroupAssignment:     resourceAppGroupAssignment(),
			appUser:                resourceAppUser(),
			appOAuth:               resourceAppOAuth(),
			appOAuthRedirectUri:    resourceAppOAuthRedirectUri(),
			appSaml:                resourceAppSaml(),
			appSecurePasswordStore: resourceAppSecurePasswordStore(),
			appSwa:                 resourceAppSwa(),
			appThreeField:          resourceAppThreeField(),
			appUserSchema:          resourceAppUserSchema(),
			appUserBaseSchema:      resourceAppUserBaseSchema(),
			authServer:             resourceAuthServer(),
			authServerClaim:        resourceAuthServerClaim(),
			authServerPolicy:       resourceAuthServerPolicy(),
			authServerPolicyRule:   resourceAuthServerPolicyRule(),
			authServerScope:        resourceAuthServerScope(),
			factor:                 resourceFactor(),
			groupRoles:             resourceGroupRoles(),
			groupRule:              resourceGroupRule(),
			idpResource:            resourceIdpOidc(),
			idpSaml:                resourceIdpSaml(),
			idpSamlKey:             resourceIdpSigningKey(),
			idpSocial:              resourceIdpSocial(),
			inlineHook:             resourceInlineHook(),
			networkZone:            resourceNetworkZone(),
			oktaGroup:              resourceGroup(),
			oktaUser:               resourceUser(),
			policyMfa:              resourcePolicyMfa(),
			policyPassword:         resourcePolicyPassword(),
			policyRuleIdpDiscovery: resourcePolicyRuleIdpDiscovery(),
			policyRuleMfa:          resourcePolicyMfaRule(),
			policyRulePassword:     resourcePolicyPasswordRule(),
			policyRuleSignOn:       resourcePolicySignonRule(),
			policySignOn:           resourcePolicySignon(),
			templateEmail:          resourceTemplateEmail(),
			trustedOrigin:          resourceTrustedOrigin(),
			userSchema:             resourceUserSchema(),
			userBaseSchema:         resourceUserBaseSchema(),

			// The day I realized I was naming stuff wrong :'-(
			"okta_idp":                       deprecateIncorrectNaming(resourceIdpOidc(), idpResource),
			"okta_saml_idp":                  deprecateIncorrectNaming(resourceIdpSaml(), idpSaml),
			"okta_saml_idp_signing_key":      deprecateIncorrectNaming(resourceIdpSigningKey(), idpSamlKey),
			"okta_social_idp":                deprecateIncorrectNaming(resourceIdpSocial(), idpSocial),
			"okta_bookmark_app":              deprecateIncorrectNaming(resourceAppBookmark(), appBookmark),
			"okta_saml_app":                  deprecateIncorrectNaming(resourceAppSaml(), appSaml),
			"okta_oauth_app":                 deprecateIncorrectNaming(resourceAppOAuth(), appOAuth),
			"okta_oauth_app_redirect_uri":    deprecateIncorrectNaming(resourceAppOAuthRedirectUri(), appOAuthRedirectUri),
			"okta_auto_login_app":            deprecateIncorrectNaming(resourceAppAutoLogin(), appAutoLogin),
			"okta_secure_password_store_app": deprecateIncorrectNaming(resourceAppSecurePasswordStore(), appSecurePasswordStore),
			"okta_three_field_app":           deprecateIncorrectNaming(resourceAppThreeField(), appThreeField),
			"okta_swa_app":                   deprecateIncorrectNaming(resourceAppSwa(), appSwa),
			"okta_password_policy":           deprecateIncorrectNaming(resourcePolicyPassword(), policyPassword),
			"okta_signon_policy":             deprecateIncorrectNaming(resourcePolicySignon(), policySignOn),
			"okta_signon_policy_rule":        deprecateIncorrectNaming(resourcePolicySignonRule(), policyRuleSignOn),
			"okta_password_policy_rule":      deprecateIncorrectNaming(resourcePolicyPasswordRule(), policyRulePassword),
			"okta_mfa_policy":                deprecateIncorrectNaming(resourcePolicyMfa(), policyMfa),
			"okta_mfa_policy_rule":           deprecateIncorrectNaming(resourcePolicyMfaRule(), policyRuleMfa),
		},
		DataSourcesMap: map[string]*schema.Resource{
			"okta_app":               dataSourceApp(),
			"okta_app_saml":          dataSourceAppSaml(),
			"okta_app_metadata_saml": dataSourceAppMetadataSaml(),
			"okta_default_policies":  deprecatedPolicies,
			"okta_default_policy":    dataSourceDefaultPolicies(),
			"okta_everyone_group":    dataSourceEveryoneGroup(),
			"okta_group":             dataSourceGroup(),
			"okta_idp_metadata_saml": dataSourceIdpMetadataSaml(),
			"okta_idp_saml":          dataSourceIdpSaml(),
			"okta_policy":            dataSourcePolicy(),
			"okta_user":              dataSourceUser(),
			"okta_users":             dataSourceUsers(),
			authServer:               dataSourceAuthServer(),
		},

		ConfigureFunc: providerConfigure,
	}
}

func deprecateIncorrectNaming(d *schema.Resource, newResource string) *schema.Resource {
	d.DeprecationMessage = fmt.Sprintf("Resource is deprecated due to a correction in naming conventions, please use %s instead.", newResource)
	return d
}

func providerConfigure(d *schema.ResourceData) (interface{}, error) {
	var err error

	log.Printf("[INFO] Initializing Okta client")
	path := d.Get("config_file").(string)

	cfg := &ini.File{}

	if path != "" {
		cfg, err = ini.Load(path)

		if err != nil {
			return nil, err
		}
	}

	config := &Config{
		orgName:     getConfigString(cfg, d, "org_name"),
		domain:      getConfigString(cfg, d, "base_url"),
		apiToken:    getConfigString(cfg, d, "api_token"),
		parallelism: getConfigInt(cfg, d, "parallelism"),
		retryCount:  getConfigInt(cfg, d, "max_retries"),
		maxWait:     getConfigInt(cfg, d, "max_wait_seconds"),
		minWait:     getConfigInt(cfg, d, "min_wait_seconds"),
		backoff:     getConfigBool(cfg, d, "backoff"),
	}

	err = config.loadAndValidate()

	return config, err
}

func getConfigString(cfg *ini.File, d *schema.ResourceData, key string) string {
	v := cfg.Section("").Key(key).String()

	if v == "" {
		return d.Get(key).(string)
	}

	return v
}

func getConfigInt(cfg *ini.File, d *schema.ResourceData, key string) int {
	v, err := cfg.Section("").Key(key).Int()

	if err != nil {
		return d.Get(key).(int)
	}

	return v
}

func getConfigBool(cfg *ini.File, d *schema.ResourceData, key string) bool {
	v, err := cfg.Section("").Key(key).Bool()

	if err != nil {
		return d.Get(key).(bool)
	}

	return v
}
